### PR TITLE
Client authentication

### PR DESCRIPTION
**Pull request by Tashima42**
Hi, I hope this PR can help you interact with the new API
## Changes
* Included a basic example of client authentication
* Integrated the login and signup page with the API
*  Added Docker to the client

To help with API and other complex interactions, I adopted a providers pattern, they are basically wrappers around the actual code, this is useful to have the default configuration centralized in one place, like the API URL, so you don't have to declare `process.env.REACT_APP...` everytime you make a request, this is also useful to change providers, if in the future you want to change the characters info API, it's really simple, just change the API URL and make some modifications to the provider, all the remaining code will still be the same.

Example:
Instead of calling: https://github.com/Abhiramborige/react-app-mahabharath/blob/7d97e37e08021810da4d90d1c82fa4ceafbea76e/src/Components/online_info.jsx#L33

You import the `charactersProvider`:
https://github.com/Abhiramborige/react-app-mahabharath/blob/a5330b2443b98b8b8f851bcefa1163b105b5741a/src/Components/online_info.jsx#L3
And use this function:
https://github.com/Abhiramborige/react-app-mahabharath/blob/a5330b2443b98b8b8f851bcefa1163b105b5741a/src/Components/online_info.jsx#L33

## Docker
With Docker and docker-compose installed, you can start all the applications with:
```
npm install
docker-compose -f docker-compose.dev.yml up
```
This will start a MongoDB container, the server container and the client container, but you still have hot reload.
If you want to learn more about Docker: https://youtu.be/Gjnup-PuquQ

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -8,8 +8,9 @@
     "@testing-library/jest-dom": "^5.11.10",
     "@testing-library/react": "^11.2.6",
     "@testing-library/user-event": "^12.8.3",
+    "axios": "^0.22.0",
     "firebase": "^8.6.8",
-    "puppeteer": "^10.1.0",
```
Comment (Tashima42) [inline package.json:12]: puppeteer can't run on the client side, this dependency was here probably by mistake

File: package.json [hunk 2]
```diff
@@ -8,8 +8,9 @@
...
+    "jwt-decode": "^3.1.2",
```
Comment (Tashima42) [inline package.json:13]: This package will be useful to access information like the user email, user id, token expiration, etc.

File: src/OtherPages/login.jsx [hunk 3]
```diff
@@ -1,28 +1,43 @@
...
+  const emailRef = React.createRef();
+  const passwordRef = React.createRef();
 
-    const _handleClick = () => {
-        return onClickCallback(emailRef.current.value, passwordRef.current.value);
+  const _handleClick = async () => {
+    const email = emailRef.current.value;
+    const password = passwordRef.current.value;
+
+    if (isLogin) {
```
Comment (Tashima42) [inline src/OtherPages/login.jsx:16]: login and signup functionality, this doesn't do much yet, but you can look on the your browser's localStorage to see user and token information.

Saving an access token on the localStorage isn't the most secure way, I'm using this by convenience, if you want to change it to a httpOnly cookie, do it.

File: src/providers/index.js [hunk 4]
```diff
@@ -0,0 +1,51 @@
+import axios from "axios"
+import jwtDecode from "jwt-decode"
+
+import buildCharacterProvider from "./character"
+import buildAuthenticationProvider from "./authentication"
+import buildStorageProvider from "./storage"
+import buildJwtProvider from "./jwt"
+
+const baseURL = "http://localhost:3050" // TODO: FIX TO USE ENV process.env.REACT_APP_MAHABHARATH_API
```
Comment (Tashima42) [inline src/providers/index.js:9]: I had some issues using the ENV variable and had to hard code the URL, if someone can debug and fix this I would be really happy.

File: src/providers/index.js [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+const http = axios.create({})
+const httpAuthenticated = axios.create({})
+
+const storageProvider = buildStorageProvider()
+const authenticationProvider = buildAuthenticationProvider({ http, baseURL, storageProvider })
```
Comment (Tashima42) [inline src/providers/index.js:15]: Here, I'm using dependency injection, the name is scary, but is a simple concept.

When using the javascript [`forEach`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach#syntax) function you can pass a `callbackFn`, this is a function that will the executed on every iteration of an array, so in JS functions can be passed as arguments to functions, and this is what we're doing here, 
```js
const http = axios.create({})
const httpAuthenticated = axios.create({})
```
here, we are creating two instances of axios, one will be used for routes who needs authentication, and the other for routes who doesn't need authentication
Now, the character information doesn't need the user to login, but if in the future you wanted to change this, you could do:
```js
const characterProvider = buildCharacterProvider({ http, baseURL }) // change this
const characterProvider = buildCharacterProvider({ http: httpAuthenticated, baseURL }) // for this
```

File: src/Components/online_info.jsx [hunk 6]
```diff
@@ -48,9 +47,11 @@ class OnlineInformation extends Component {
   }
 
   render() {
+    /*
```
Comment (Abhiramborige) [inline src/Components/online_info.jsx:50]: why did you comment out this timeout error, i think this is necessary when fetch fails.

File: src/OtherPages/login.jsx (hunk 3 above)
Comment (Abhiramborige) [reply src/OtherPages/login.jsx:16]: any reference to learn this concept ?

File: src/providers/index.js (hunk 5 above)
Comment (Abhiramborige) [reply src/providers/index.js:15]: yeah got the idea.

File: Dockerfile.dev [hunk 7]
```diff
@@ -0,0 +1,4 @@
+FROM node:alpine
```
Comment (Abhiramborige) [inline Dockerfile.dev:1]: I don't have any idea about this docker stuff, what exactly it is, and where to learn this.

File: src/Components/online_info.jsx (hunk 6 above)
Comment (Tashima42) [reply src/Components/online_info.jsx:50]: This is a good video to understand the concept: https://youtu.be/Gjnup-PuquQ

File: Dockerfile.dev (hunk 7 above)
Comment (Tashima42) [reply Dockerfile.dev:1]: This is a good video to understand the concept: https://youtu.be/Gjnup-PuquQ

File: src/Components/online_info.jsx (hunk 6 above)
Comment (Tashima42) [reply src/Components/online_info.jsx:50]: Just throwing an error is not a good way to do error treatment, this just breaks the application, maybe you could present an alert to the user temporarily, before adding real error treatment.

File: src/OtherPages/login.jsx (hunk 3 above)
Comment (Tashima42) [reply src/OtherPages/login.jsx:16]: You can learn something from this post: https://www.bezkoder.com/react-refresh-token/
But seeing the nature of the application, saving the access token on the local storage is totally ok, there nothing much an attacker could even if he could hijack this token.

File: src/Components/online_info.jsx (hunk 6 above)
Comment (Abhiramborige) [reply src/Components/online_info.jsx:50]: Yeah, but I am not getting the concept of how to remove the error that is enclosed by the ErrorBoundary component in react to catch that error

Comment (Abhiramborige) [reply src/Components/online_info.jsx:50]: I also want to continue the the application after the error is thrown by just rendering a button for trying to refetch the data by clicking on it. Here we can reset the error by clicking the button and then try to fetch same data again. But I am not getting how to implement that using errorboundaries in react. Could you please help me.

**Comment by Abhiramborige** [comment]
Thanks for the contribution 😄

File: src/Components/online_info.jsx (hunk 6 above)
Comment (Tashima42) [reply src/Components/online_info.jsx:50]: You probably should use React State Hook for this: https://reactjs.org/docs/hooks-state.html
I don't have much experience in frontend development, so I don't know exactly how you should implement it, but using the application state is probably the best approach.

**Pull request merged**